core/src/main/java/hudson/model/Descriptor.java

```diff
@@ -909,14 +909,19 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
         if (formData!=null) {
             for (Object o : JSONArray.fromObject(formData)) {
                 JSONObject jo = (JSONObject)o;
-                String kind = jo.optString("$class", null);
-                if (kind == null) {
-                  // Legacy: Remove once plugins have been staged onto $class
-                  kind = jo.getString("kind");
+                Descriptor<T> d = null;
+                String kind = jo.optString("kind", null);
+                if (kind != null) {
+                    d = find(descriptors, kind);
```

Comment:
I wonder it would be cleaner to introduce new "FindById" and "findByClassName" methods and explicitly invoke one of them. find would then use both and marked as deprecated.